app/scripts/services/planner.service.js

```diff
@@ -66,18 +66,34 @@
             // check to see if we've finished all the bosses
             if (step.bosses) {
                 angular.forEach(step.bosses, function(boss) {
-                    if (boss.isAlliance && items.isAlliance && items.lookup[boss.ID] === undefined) {
-                        neededBosses.push(boss);
-                        completed = false;
+
+                    var bossIsNeutral = !boss.isAlliance && !boss.isHorde,
+                        character = items, // aliasing for clarity
+                        characterNeedsBoss = function(id){ return !character.lookup[id]; },
+                        addBoss = function(boss) {
+                            neededBosses.push(boss);
+                            completed = false;
+                        };
+
+                    if (showAll) { addBoss(boss); return; }
+                    if (boss.ID === undefined) { return; } // continue the loop, bad boss data
+                    if (!characterNeedsBoss(boss.ID)) { return; }
+
+                    if (bossIsNeutral) {
+                        addBoss(boss);
+                        return;
                     }
-                    else if (boss.isHorde && !items.isAlliance && items.lookup[boss.ID] === undefined) {
-                        neededBosses.push(boss);
-                        completed = false;
+
+                    if (boss.isAlliance && character.isAlliance) {
+                        addBoss(boss);
+                        return;
                     }
-                    else if ((boss.ID !== undefined && items.lookup[boss.ID] === undefined) || showAll) {
-                        neededBosses.push(boss);
-                        completed = false;
+
+                    if (boss.isHorde && character.isHorde) {
+                        addBoss(boss);
+                        return;
                     }
```

Comment:
Since all those blocks are the same, wouldn't it make more sense to have just one big condition which enters the block?

Reply:
Sure, I had it this way to enhance condition readability. I made the change, let me know what you think.